### PR TITLE
Fix some pub use privacy errors and a bit of resolve refactoring

### PR DESCRIPTION
**Pull request by nrc**
r? @eddyb

### CODE REVIEW

File: src/libsyntax/test.rs [hunk 1]
```diff
@@ -142,9 +140,18 @@ impl<'a> fold::Folder for TestHarnessGenerator<'a> {
                     self.tests.push(i.ident);
                     // debug!("have {} test/bench functions",
                     //        cx.testfns.len());
+
+                    // Make all tests public so we can call them from outside
+                    // the module (note that the tests are re-exported and must
+                    // be made public themselves to avoid privacy errors).
+                    let mut result = (*i).clone();
```
Comment (eddyb) [inline src/libsyntax/test.rs]: You're never supposed to clone AST nodes if you own them (can we make specific impls of `Clone` warn/error?), here's what you can do instead:
```rust
i.map(|mut i| { i.vis = ast::Public; i })
```

**Comment by eddyb** [comment]
r=me with nits fixed - are there really no issues closed by this?

**Comment by nrc** [comment]
#23266 is closed by this (see the comment on the second commit)

**Pull request merged**